### PR TITLE
Support both sequential and parallel execution strategies for running synchronization

### PR DESCRIPTION
**Pull request by kabirbaidhya**
**Support sequential execution strategy with a configurable option** 
- Add a new config option `execution` supporting two choices `sequential` and `parallel` - default is still parallel execution so no breaking change.
- Update utilities and add tests:
   - Change `promise.runSequentially` to accept promisers instead of promises.
   - Extract the execution process to `services/execution.ts` service.
   - Add tests
- This is an opt-in option for the user. If they want to run it in a sequential mode (one connection at a time), they'll need to add this to their config:
   ```
   execution: sequential
   ```

### Preview
**sequential**
sequential-run

**parallel**
parallel-run

### CODE REVIEW

File: src/constants.ts [hunk 1]
```diff
@@ -16,6 +16,7 @@ export const CONNECTIONS_FILENAME = 'connections.sync-db.json';
 export const INJECTED_CONFIG_TABLE = '__sync_db_injected_config';
 export const DEFAULT_CONFIG: SyncConfig = {
   basePath: path.resolve(process.cwd(), 'src/sql'),
+  executionStrategy: 'parallel',
```
Comment (kabirbaidhya) [inline src/constants.ts]: Could `execution` be a better name here?

File: src/services/execution.ts [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+  switch (config.execution) {
+    case 'sequential':
+      return runSequentially(processes);
+
+    case 'parallel':
+      return Promise.all(processes.map(fn => fn()));
+
+    default:
+      throw new Error(`Invalid execution strategy found in the configuration "${config.execution}".`);
```
Comment (mesaugat) [inline src/services/execution.ts]:
```suggestion
      throw new Error(`Execution strategy should be "sequential" or "parallel" found: "${config.execution}".`);
```

File: test/service/execution.test.ts [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+      expect(result).to.deep.equal(['Task A', 'Task B', 'Task C', 'Task D']);
+      expect(tracker).to.deep.equal(['Task C', 'Task D', 'Task A', 'Task B']);
+    });
+
+    it('should throw an error if unknown strategy is provided.', () => {
+      const processes = getProcesses([]);
+
+      expect(() => executionService.executeProcesses(processes, { execution: 'future' } as any)).to.throw(
+        'Invalid execution strategy found in the configuration "future".'
```
Comment (mesaugat) [inline test/service/execution.test.ts]: .

File: test/service/execution.test.ts [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+      .then(() => 'Task C'),
+  () =>
+    timeout(40)
+      .then(() => tracker.push('Task D'))
+      .then(() => 'Task D')
+];
+
+describe('Services: execution', () => {
+  describe('executeProcesses', () => {
+    it('should be able to run things with sequentially [strategy = sequential].', async () => {
```
Comment (mesaugat) [inline test/service/execution.test.ts]:
```suggestion
    it('should run sequentially with [strategy = sequential].', async () => {
```

File: test/service/execution.test.ts [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+      const result = await executionService.executeProcesses(processes, {
+        execution: 'sequential'
+      } as SyncConfig);
+
+      expect(result).to.deep.equal(['Task A', 'Task B', 'Task C', 'Task D']);
+      expect(tracker).to.deep.equal(['Task A', 'Task B', 'Task C', 'Task D']);
+    });
+
+    it('should be able to run things in-parallel [strategy = parallel].', async () => {
```
Comment (mesaugat) [inline test/service/execution.test.ts]:
```suggestion
    it('should run in-parallel with [strategy = parallel].', async () => {
```

File: test/service/execution.test.ts [hunk 6]
```diff
@@ -0,0 +1,61 @@
+import { expect } from 'chai';
+import { it, describe } from 'mocha';
+
+import { timeout } from '../../src/util/promise';
+import SyncConfig from '../../src/domain/SyncConfig';
+import * as executionService from '../../src/services/execution';
+
+const getProcesses = (tracker: string[]) => [
+  () =>
+    timeout(45)
```
Comment (mesaugat) [inline test/service/execution.test.ts]: Add a test where the promiser rejects something?

Comment (kabirbaidhya) [reply test/service/execution.test.ts]: Done.

**Pull request merged**